CNV-20188: Mass machine type change procedure (#91316)
openshift/openshift-docs: modules/virt-updating-multiple-vms.adoc

// Module included in the following assemblies:
//
// * virt/virtual_machines/virt-edit-vms.adoc

:_mod-docs-content-type: PROCEDURE
[id="virt-updating-multiple-vms_{context}"]
= Updating multiple virtual machines

> You can update multiple virtual machines (VMs) at the same time by using the command line.

Reviewer: ```suggestion
You can use the command-line interface (CLI) to update multiple virtual machines (VMs) at the same time.
```

![Screenshot From 2025-05-08 14-43-39](https://github.com/user-attachments/assets/d2d5101f-d715-4984-af54-e095088beab8)

CLI is mentioned in the prereq so might be good to go with command-line interface.

![Screenshot From 2025-05-08 14-43-39](https://github.com/user-attachments/assets/570f93b3-7bde-4032-a5ab-aecdcb228cbf)

Guidance is not 100 % clearly, IMO, but the CLI is referenced in the prereq section.

.Prerequisites

* You installed the `oc` CLI.

> * You have access to the {product-title} cluster with `cluster-admin` permissions.

Reviewer: ```suggestion
* You have access to a {product-title} cluster and you have `cluster-admin` permissions.
```

.Procedure

> . Create a privileged service account by running the following commands:

Reviewer: More context for what each command might be useful to give a customer onus on the importance of the sequence of commands.

+
[source,terminal]
----
$ oc adm new-project kubevirt-api-lifecycle-automation
----
+
[source,terminal]
----
$ oc create sa kubevirt-api-lifecycle-automation -n kubevirt-api-lifecycle-automation
----
+
[source,terminal]
----
$ oc create clusterrolebinding kubevirt-api-lifecycle-automation --clusterrole=cluster-admin --serviceaccount=kubevirt-api-lifecycle-automation:kubevirt-api-lifecycle-automation
----

. Determine the pull URL for the `kubevirt-api-lifecycle` image by running the following command:
+
[source,terminal]
----
$ oc get csv -n openshift-cnv -l=operators.coreos.com/kubevirt-hyperconverged.openshift-cnv -ojson | jq '.items[0].spec.relatedImages[] | select(.name|test(".*kubevirt-api-lifecycle-automation.*")) | .image'
----

> . Deploy `Kubevirt-Api-Lifecycle-Automation` by creating a job object as shown in the following example:

Reviewer 2: For another time:
```suggestion
. Deploy the `Kubevirt-Api-Lifecycle-Automation` job by creating a job object as shown in the following example:
```

+
[source,yaml]
----
apiVersion: batch/v1
kind: Job
metadata:
 name: kubevirt-api-lifecycle-automation
 namespace: kubevirt-api-lifecycle-automation
spec:
 template:
  spec:
   containers:
   - name: kubevirt-api-lifecycle-automation
     image: quay.io/openshift-virtualization/kubevirt-api-lifecycle-automation:v4.19 <1>
     imagePullPolicy: Always
     env:
     - name: MACHINE_TYPE_GLOB <2>
       value: smth-glob9.10.0
     - name: RESTART_REQUIRED <3>
       value: "true"
     - name: NAMESPACE <4>
       value: "default"
     - name: LABEL_SELECTOR <5>
       value: my-vm
     securityContext:
      allowPrivilegeEscalation: false
      capabilities:
       drop:
       - ALL
      privileged: false
      runAsNonRoot: true
      seccompProfile:
       type: RuntimeDefault
   restartPolicy: Never
   serviceAccountName: kubevirt-api-lifecycle-automation
----

> <1> Replace the image value with the image URL from the previous step.

Reviewer: If a new step is added in the future, "previous" might be confusing here. 

> Replace the image value with your pull URL for the image.

<2> Replace the `MACHINE_TYPE_GLOB` value with your own pattern. This pattern is used to detect deprecated machine types that need to be upgraded.

> <3> If the `RESTART_REQUIRED` value is set to `true`, VMs are restarted after the machine type is updated. If you do not want them to be restarted, set the value to `false`.

Reviewer: ```suggestion
<3> If the `RESTART_REQUIRED` emvironment variable is set to `true`, VMs are restarted after the machine type is updated. If you do not want VMs to be restarted, set the value to `false`.
```

> <4> The `namespace` environment value indicates the namespace to look for VMs in. Leave it empty for the job to go over all namespaces in the cluster.

Reviewer: 4> The `namespace` environment value indicates the namespace to look for VMs in. Leave the parameter empty for the job to go over all namespaces in the cluster.

> <5> You can use the `LABEL_SELECTOR` environment value for selecting VMs to apply the job on. If you want the job to go over all VMs in the cluster, leave the value empty.

Reviewer: ```suggestion
<5> You can use the `LABEL_SELECTOR` environment variable to select VMs that receive the job action. If you want the job to go over all VMs in the cluster, do not assign a value for the parameter.
```

> <4> The `namespace` environment value indicates the namespace to look for VMs in. Leave the parameter empty for the job to go over all namespaces in the cluster.
> <5> You can use the `LABEL_SELECTOR` environment variable to select VMs that receive the job action. If you want the job to go over all VMs in the cluster, do not assign a value to the parameter.

Reviewer 2: I'm not sure _go over_ is sufficient here. A bit slang to me.
```suggestion
<4> The `namespace` environment value indicates the namespace to look for VMs in. Leave the parameter empty for the job to look in all namespaces in the cluster.
<5> You can use the `LABEL_SELECTOR` environment variable to select VMs that receive the job action. If you want the job to look in all VMs in the cluster, do not assign a value to the parameter.
```

// Module included in the following assemblies:
//
// * virt/virtual_machines/virt-edit-vms.adoc

:_mod-docs-content-type: PROCEDURE
[id="virt-updating-multiple-vms_{context}"]
= Updating multiple virtual machines

You can update multiple virtual machines (VMs) at the same time by using the command line.

.Prerequisites

* You installed the `oc` CLI.
* You have access to the {product-title} cluster with `cluster-admin` permissions.

.Procedure

. Create a privileged service account by running the following commands:
+
[source,terminal]
----
$ oc adm new-project kubevirt-api-lifecycle-automation
----
+
[source,terminal]
----
$ oc create sa kubevirt-api-lifecycle-automation -n kubevirt-api-lifecycle-automation
----
+
[source,terminal]
----
$ oc create clusterrolebinding kubevirt-api-lifecycle-automation --clusterrole=cluster-admin --serviceaccount=kubevirt-api-lifecycle-automation:kubevirt-api-lifecycle-automation
----
+
[source,terminal]
----

> $ oc adm policy add-scc-to-user privileged -z kubevirt-api-lifecycle-automation

Reviewer 3: @dasionov i'm not sure this is needed, did you try making it work without it?

Reviewer 4: no unfortunately

Reviewer 3: I think we can delete this for now

----

. Determine the pull URL for the `kubevirt-api-lifecycle` image by running the following command:
+
[source,terminal]
----
$ oc get csv -n openshift-cnv -l=operators.coreos.com/kubevirt-hyperconverged.openshift-cnv -ojson | jq '.items[0].spec.relatedImages[] | select(.name|test(".*kubevirt-api-lifecycle-automation.*")) | .image'
----

. Deploy `Kubevirt-Api-Lifecycle-Automation` by creating a job object as shown in the following example:
+
[source,yaml]
----
apiVersion: batch/v1
kind: Job
metadata:
name: kubevirt-api-lifecycle-automation
namespace: kubevirt-api-lifecycle-automation
spec:
template:
spec:
containers:
- name: kubevirt-api-lifecycle-automation
image: quay.io/openshift-virtualization/kubevirt-api-lifecycle-automation:v4.19 <1>
imagePullPolicy: Always
env:
- name: MACHINE_TYPE_GLOB <2>
value: smth-glob9.10.0
- name: RESTART_REQUIRED <3>
value: "true"
- name: NAMESPACE <4>
value: "default"
 - name: LABEL_SELECTOR <5>

> value: my-vm

Reviewer 3: Not sure if we should add these to the example. Maybe we can put them as comments? The manifests look better as a default without them.

Author: @Barakmor1 
Do you mean the values?

Reviewer 3: After some additional thinking the current state should be enough

securityContext:
allowPrivilegeEscalation: false
capabilities:
drop:
- ALL
privileged: false
runAsNonRoot: true
seccompProfile:
type: RuntimeDefault
restartPolicy: Never
serviceAccountName: kubevirt-api-lifecycle-automation

> ----

Reviewer 3: It looks like there might be an issue with the indentation in the manifest.

Author: @Barakmor1 
Please check the indentation.
